### ISSUE TITLE
Need help with decoder training

### ISSUE THREAD

**Issue by xiankgx** [lucidrains/DALLE2-pytorch#40]
I'm training on the CC3M. Is there anything wrong with my training? The loss seems to be going down way too fast and despite the low training loss values, sampling doesn't seem to show it is working. Sampling during training by just calling decoder.sample() giving it the CLIP image embeddings of the minibatch training images. Since I'm training a decoder with two Unets and just training the first Unet for now, I'm breaking out after sampling from the first Unet.

![decoder_training_loss](https://user-images.githubusercontent.com/4113258/166122595-d619cc30-fbf5-4e87-9ae0-e1da25863592.jpg)

Theses are the samples at the 0k, 5k, 13k, 16k, and 17k training steps.

![0k](https://user-images.githubusercontent.com/4113258/166122631-39c908da-cebc-4ab3-863e-75a1292d0329.png)
![5k](https://user-images.githubusercontent.com/4113258/166122604-9aca3ea8-aa8d-4963-a2e0-0523f341d23a.png)
![13k](https://user-images.githubusercontent.com/4113258/166122608-7f3a5d2e-ec44-4485-9c54-98e260e8d30a.png)
![16k](https://user-images.githubusercontent.com/4113258/166122612-cd98526e-dc82-4aba-b660-cc1b9e1a5742.png)
![17k](https://user-images.githubusercontent.com/4113258/166122599-cf52c6a7-b0fc-41e9-a249-333580969c46.png)

**Comment by lucidrains** [comment]
@xiankgx thank you for testing it out! so based on my experience with DDPMs, the loss has to hit around 0.05 before images come into view. how large are your batch sizes?

**Comment by lucidrains** [comment]
@xiankgx also, i have yet to code this up, but ideally during sampling, we use the exponentially moving averaged unets https://github.com/lucidrains/DALLE2-pytorch/blob/main/dalle2_pytorch/train.py#L94

**Comment by xiankgx** [comment]
> @xiankgx thank you for testing it out! so based on my experience with DDPMs, the loss has to hit around 0.05 before sampled images come into view. how large are your batch sizes?
> 
> i'll be testing out the decoder training too, probably tomorrow evening-ish, as i'm planning to copy the cascading unet into the other ddpm repository i have (and test out training on oxford flowers dataset)

Batch size is 128, precision at fp32. I guess I was just not patient enough?

**Comment by xiankgx** [comment]
> @xiankgx also, i have yet to code this up, but ideally during sampling, we use the exponentially moving averaged unets https://github.com/lucidrains/DALLE2-pytorch/blob/main/dalle2_pytorch/train.py#L94 still thinking about how to architect this

I think we can still use the non-EMA one for testing?

**Comment by lucidrains** [comment]
> > @xiankgx thank you for testing it out! so based on my experience with DDPMs, the loss has to hit around 0.05 before sampled images come into view. how large are your batch sizes?
> > i'll be testing out the decoder training too, probably tomorrow evening-ish, as i'm planning to copy the cascading unet into the other ddpm repository i have (and test out training on oxford flowers dataset)
> 
> Batch size is 128, precision at fp32. I guess I was just not patient enough?

hmm, that sounds sufficient, but also very likely a bug or two remaining 🐞  that's the risk of being the first one to try a new repository 🤣

**Comment by lucidrains** [comment]
@xiankgx let me get the EMA-ed sampling code in there, just thought of a way to do it cleanly

**Comment by lucidrains** [comment]
https://github.com/lucidrains/DALLE2-pytorch/commit/ebe01749ed0a48aa77e236eb609440db0944eada tada

**Comment by xiankgx** [comment]
This training btw is only until before all the changes in the cond_drop_prob changes in https://github.com/lucidrains/DALLE2-pytorch/issues/38. 

You mentioned that text encodings must be padded to max length, and I see you are padding the text encodings in the code.
Do you think it affects the training, if I'm just padding text_encodings at the dataset/dataloader level, and I'm not really conditioning on text_encodings anyway.

**Comment by lucidrains** [comment]
i think its probably worth leaving out the text encoding conditioning altogether, and the image embedding conditioning too (with a conditional dropout of 1.) until we see that the cascading DDPM can generate things unconditionally (which it should). then i would ascertain image embedding condition works, followed by validating that classifier free guidance improves things later in training (it probably wouldn't do much early on). i would save the text encoding for last, since i think this is not the novelty of the paper and was present even in GLIDE with poor results

**Comment by lucidrains** [comment]
@xiankgx i've added the ability to shortcut sampling at a certain unet in the cascade btw, just so you have an easier time https://github.com/lucidrains/DALLE2-pytorch/commit/8260fc933a9a5118e18f209314ecfad246a42454 thanks for being the first guinea pig lol

**Comment by xiankgx** [comment]
> @xiankgx i've added the ability to shortcut sampling at a certain unet in the cascade btw, just so you have an easier time [8260fc9](https://github.com/lucidrains/DALLE2-pytorch/commit/8260fc933a9a5118e18f209314ecfad246a42454) thanks for being the first guinea pig lol

Thank you, ya, this is exactly what I had in my code.

**Comment by lucidrains** [comment]
@xiankgx i wouldn't harp on the text encodings - as shown in the paper, DALLE-2 still had trouble not treating the text as a bag of words (say binding the wrong color to a certain object). it will probably be a continued research area to see how to improve it from what they have

**Comment by lucidrains** [comment]
@xiankgx the most important thing for me is to validate whether this prior network is necessary or some distraction (whether the cascading DDPM is making things so pretty users are forgetting about its limitations)

**Comment by xiankgx** [comment]
This is at 39k iterations, top images is sampled. Bottom images are training images. I'm seeing the word "alamy" in the bottom left image in the generated images, in line with the same image from training images. But I'm still seeing quite alot of plain colored backgrounds.

![39k](https://user-images.githubusercontent.com/4113258/166150411-d08b663a-6dbe-4c40-969a-b900874e8043.png)
![39k_data](https://user-images.githubusercontent.com/4113258/166150416-9f17302e-ea97-4552-af0b-5cc793092eaa.png)

**Comment by lucidrains** [comment]
@xiankgx when you save the image, are you unnormalizing the image? (back to range of 0 to 1)?

**Comment by lucidrains** [comment]
otherwise, the general shapes and structure looks about right for DDPM-type training!

**Comment by xiankgx** [comment]
Ya, I'm doing inverse normalization with (t * std) + mean with mean and std both (0.5, 0.5, 0.5).

**Comment by xiankgx** [comment]
It depends on which specific step. Some steps are still like earlier samples where we have all images orange, red, blue. And all samples from the same step have the same color. Sometimes it is better like the above.

**Comment by lucidrains** [comment]
@xiankgx ohh got it, it would be nice to see what the text is too, if you are encoding on that. maybe its latched onto some color in the sentence, if present

**Comment by lucidrains** [comment]
what is "alamy"? is it a watermark?

**Comment by xiankgx** [comment]
Ya, I think it is a watermark. alamy should be some stock photo site.

**Comment by xiankgx** [comment]
42k, 44k, 49k. Alot of them are like this.

![42k](https://user-images.githubusercontent.com/4113258/166151425-3b6035e2-c90e-406e-82a3-850dba9cfc63.png)
![44k](https://user-images.githubusercontent.com/4113258/166151429-8b33b6f1-1494-412d-a43f-a68ee10a60d1.png)
![49k](https://user-images.githubusercontent.com/4113258/166151430-1a6af248-fa99-4042-834f-aa0f720ccddd.png)

**Comment by lucidrains** [comment]
when sampling, is each image being conditioned on a different text, or all the same?

**Comment by xiankgx** [comment]
The model is not conditioned on text. It is conditioning on CLIP image embeddings obtained from training images in the same step. It is not fixed.

**Comment by lucidrains** [comment]
yea, i think the exponential moving average should help, though what you are seeing is still a bit extreme. hoping there is not a bug in the unet lol

**Comment by lucidrains** [comment]
@xiankgx got it, i'm not sure what's going on then. do you have the sparse attention turned off? (with the grid attention) that was more of an experimental feature i have yet to test locally

**Comment by lucidrains** [comment]
@xiankgx i'll join you soon next week, as i port the cascading DDPM over and start doing local experiments on my machine with oxford dataset

**Comment by xiankgx** [comment]
> @xiankgx got it, i'm not sure what's going on then. do you have the sparse attention turned off? (with the grid attention) that was more of an experimental feature i have yet to test locally

Ya, I'm just using the defaults.

**Comment by lucidrains** [comment]
@xiankgx ok, i have it turned off

i'm not sure what's going on! your batch sizes seem large enough. i guess if you start an exponential moving average run and still see the weird color shifting, i'll be certain there's some remaining bug to squash from the decoder

**Comment by xiankgx** [comment]
![openai_1k](https://user-images.githubusercontent.com/4113258/166151885-2c2030fc-2292-4f17-9932-596338a33027.png)

**Comment by xiankgx** [comment]
Meanwhile, I am also piecing together stuffs from openai DDPM and GLIDE repos. This is with just 1k of training.

**Comment by xiankgx** [comment]
64x64 resolution. I've looked thru the code a few thousand times and I can't find anything as well.

**Comment by lucidrains** [comment]
@xiankgx ahh ok, i think i spot the potential problem. so i changed the conditioning in the U-net over to a cross-attention based, but its probably best to leave the time conditioning out of it - let me offer an additional time conditioning the old fashioned way (MLP + sum to hiddens)

**Comment by xiankgx** [comment]
In the other version, I'm doing both, proj image embed and add to time conditioning or embedding, and proj image embed as 4 tokens and concat with text sequence.

**Comment by lucidrains** [comment]
@xiankgx it doesn't make sense for time to have to fight for attention with the text. you always need to factor in time

**Comment by lucidrains** [comment]
@xiankgx ohh no way, and still seeing the same thing? ok, i've got nothing then

you should try the exponential moving average run, that's the only thing i can think of

**Comment by lucidrains** [comment]
@xiankgx let me add an extra conditioning for time for this sunday morning, since it just makes more sense

**Comment by lucidrains** [comment]
@xiankgx ok, should be all there, and in addition, may have spotted an error with the initial convnext block (it was actually conditioning in pixel space) i've added an extra initial convolution to make sure that doesn't happen

**Comment by xiankgx** [comment]
> sigh, it looks like batchnorm and groupnorms will have to stick around for a bit longer 😢

What about them? You have new findings on this problem? I've stopped training meanwhile.

**Comment by lucidrains** [comment]
@xiankgx yes, it was the convnext blocks, jumped the gun in adopting it. it is better to stick with resnet blocks for generative work until some other paper proves it out. welcome you to retry if you have the time! :pray: otherwise i'll get into the weeds this weekend

**Comment by chinoll** [comment]
I have encountered the same problem as you. Have you solved it?

**Comment by xiankgx** [comment]
The run was run when the code was in its earlier stage when ConvNext blocks was used instead of ResNet blocks. I suspected it could be a problem somewhere in the U-Net causing entire batches to have the same tint or color. I have not retried training with the newer ResNet blocks. Is your U-Net using the newer code with ResNet blocks?

**Comment by egeozsoy** [comment]
I am training the Decoder with two Unets with DecoderTraining, using the automated mixed precision. My image size is 224x224. I am only managing a batch size of 16 on 12GB VRAM. I was wondering if this is normal as @xiankgx was using a batch size of 128 while working with float32.

**Comment by xiankgx** [comment]
> I am training the Decoder with two Unets with DecoderTraining(described in README.MD), using automated mixed precision. My image size is 224x224. I am only managing a batch size of 16 on 12GB VRAM. I was wondering if this is normal as @xiankgx was using a batch size of 128 while working with float32.

Hehe, I was running on 4 GPUs with 40 GB RAM each. No worries on that.

**Comment by lucidrains** [comment]
![sample-5 0](https://user-images.githubusercontent.com/108653/168661528-0aa85439-8581-4011-b419-8953999420ad.png)

i do believe the decoder trainer is working! oxford flowers dataset at just 5k

**Comment by lucidrains** [comment]
also think the cross embed layers from the Crossformer paper is leading to faster convergence (at least, in comparison to https://github.com/lucidrains/denoising-diffusion-pytorch with a quick eyeball)

**Comment by lucidrains** [comment]
![sample-15 0](https://user-images.githubusercontent.com/108653/168672964-a30d5b6b-caba-40d7-a484-dea97ee8a562.png)

7k steps, yea i think it is working :) @xiankgx your issue is likely from the convnext blocks + lack of EMA

**Comment by lucidrains** [comment]
@egeozsoy are you seeing any results? if you confirm it is working on your end, i'm closing this issue

**Comment by lucidrains** [comment]
so much easier than GAN training, thank the lord this technique supplanted adversarial methods

**Comment by egeozsoy** [comment]
Yes, I am also getting some results. Though for me the older learning rate (3e-4) was working better I think. Just fyi

**Comment by lucidrains** [comment]
@egeozsoy yup, i'll change it to what they had in the paper, which is around `1e-4` thank you for confirming!
